packages/zoe/src/contracts/helpers/bondingCurves.js

```diff
@@ -1,102 +1,82 @@
-          subtract(yReserve, yExtentOut),
-        );
-        return {
-          amountOut: yAmountMath.make(yExtentOut),
-          newPoolAmountsArray: harden(newPoolAmountsArray),
-        };
-      }
-      if (brandIn === brands[Y]) {
-        const yExtentIn = amountIn.extent;
-        const xExtentOut = getInputPrice(yExtentIn, yReserve, xReserve);
-        const newPoolAmountsArray = [...poolAmountsArray];
-        newPoolAmountsArray[X] = xAmountMath.make(
-          subtract(xReserve, xExtentOut),
-        );
-        newPoolAmountsArray[Y] = yAmountMath.make(add(yReserve, yExtentIn));
-        return {
-          amountOut: xAmountMath.make(xExtentOut),
-          newPoolAmountsArray: harden(newPoolAmountsArray),
-        };
-      }
-      throw new Error(`amountIn ${amountIn} were malformed`);
+      const outputExtent = floorDivide(numerator, denominator);
+      const newOutputReserve = subtract(outputReserve, outputExtent);
+      const newInputReserve = add(inputReserve, inputExtent);
+      return harden({ outputExtent, newInputReserve, newOutputReserve });
     },
 
     // Calculate how many liquidity tokens we should be minting to
     // send back to the user when adding liquidity. Calculations are
     // based on the extents represented by index 0. If the current
     // supply is zero, start off by just taking the extent at index 0
     // and using it as the extent for the liquidity token.
-    calcLiqExtentToMint: (liqTokenSupply, poolAmounts, userAmounts) =>
+    calcLiqExtentToMint: ({ liqTokenSupply, inputExtent, inputReserve }) =>
       liqTokenSupply > 0
-        ? floorDivide(
-            multiply(userAmounts[0].extent, liqTokenSupply),
-            poolAmounts[0].extent,
-          )
-        : userAmounts[0].extent,
+        ? floorDivide(multiply(inputExtent, liqTokenSupply), inputReserve)
+        : inputExtent,
 
     // Calculate how many underlying tokens (in the form of amount)
     // should be returned when removing liquidity.
-    calcAmountsToRemove: (liqTokenSupply, poolAmounts, liquidityAmountsIn) =>
-      poolAmounts.map((amount, i) =>
-        amountMathArray[i].make(
+    calcAmountsToRemove: ({
+      liqTokenSupply,
+      poolAmounts,
+      liquidityExtentIn,
+    }) => {
+      const { issuerKeywordRecord } = zoe.getInstanceRecord();
+      const amountMaths = zoe.getAmountMaths(issuerKeywordRecord);
+      const newUserAmounts = harden({
+        TokenA: amountMaths.TokenA.make(
+          floorDivide(
+            multiply(liquidityExtentIn, poolAmounts.TokenA.extent),
+            liqTokenSupply,
+          ),
+        ),
+        TokenB: amountMaths.TokenB.make(
           floorDivide(
-            multiply(liquidityAmountsIn.extent, amount.extent),
+            multiply(liquidityExtentIn, poolAmounts.TokenB.extent),
             liqTokenSupply,
           ),
         ),
-      ),
+        Liquidity: amountMaths.Liquidity.getEmpty(),
```

Comment:
Should this line be omitted?

Reply:
No, we cannot yet reallocate for a subset of keywords